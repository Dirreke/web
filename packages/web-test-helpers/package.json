{
  "name": "web-test-helpers",
  "private": true,
  "peerDependencies": {
      "@vue/composition-api": "1.7.1",
      "vue-gettext": "^2.1.5",
      "vue-router": "3.6.5",
      "vuex": "^3.6.2"
  }
}
