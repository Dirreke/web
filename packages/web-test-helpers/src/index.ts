export * from './mocks/defaultComponentMocks'
export * from './mocks/defaultStubs'
export * from './mocks/axios'

export * from './defaultPlugins'
export * from './helpers'
