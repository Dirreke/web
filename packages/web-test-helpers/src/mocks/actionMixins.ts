export const getActionMixinMocks = ({
  actions,
  enabledActions = [],
  additionalActions = undefined
}: {
  actions: string[]
  enabledActions?: string[]
  additionalActions?: { [key: string]: any }
}) => {
  let mocks = {}
  if (additionalActions) {
    mocks = { ...additionalActions }
  }
  for (const action of actions) {
    const isEnabled = !!enabledActions.includes(action)
    mocks[action] = [{ isEnabled: () => isEnabled, name: '', items: [] }]
  }
  return mocks
}
