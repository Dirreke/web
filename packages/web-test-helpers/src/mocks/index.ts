export * from './actionMixins'
export * from './axios'
