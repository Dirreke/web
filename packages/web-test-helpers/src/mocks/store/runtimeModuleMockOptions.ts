export const runtimeModuleMockOptions = {
  runtime: {
    namespaced: true,
    modules: {
      spaces: {
        namespaced: true,
        getters: {
          spaces: jest.fn(() => [])
        },
        mutations: {
          UPDATE_SPACE_FIELD: jest.fn()
        },
        actions: {
          loadSpaceMembers: jest.fn(),
          reloadProjectSpaces: jest.fn()
        }
      }
    }
  }
}
