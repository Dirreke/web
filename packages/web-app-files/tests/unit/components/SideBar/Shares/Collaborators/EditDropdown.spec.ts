import { Settings } from 'luxon'
import EditDropdown from 'web-app-files/src/components/SideBar/Shares/Collaborators/EditDropdown.vue'
import { defaultStoreMockOptions } from 'web-test-helpers/src/mocks/store/defaultStoreMockOptions'
import { defaultComponentMocks } from 'web-test-helpers/src/mocks/defaultComponentMocks'
import { mock } from 'jest-mock-extended'
import { createStore, defaultPlugins, mount } from 'web-test-helpers'

const selectors = {
  actionMenuItem: '.action-menu-item',
  showDetailsBtn: '.show-access-details',
  removeShareBtn: '.remove-share',
  findCollaboratorsExpirationBtn: '.files-collaborators-expiration-button',
  removeExpirationDateBtn: '.remove-expiration-date',
  datepickerStub: 'oc-datepicker-stub'
}

describe('EditDropdown', () => {
  beforeEach(() => {
    Settings.defaultZone = 'utc'
  })
  describe('default actions', () => {
    it('render "remove share" and "access details" as default', () => {
      const { wrapper } = createWrapper()
      expect(wrapper.findAll(selectors.actionMenuItem).length).toBe(2)
      expect(wrapper).toMatchSnapshot()
    })
    it('call the "showAccessDetails"-action on click', async () => {
      const { wrapper } = createWrapper()
      const emitSpy = jest.spyOn(wrapper.vm, '$emit')
      await wrapper.findAll(selectors.showDetailsBtn).trigger('click')
      expect(emitSpy).toHaveBeenCalledWith('showAccessDetails')
    })
    it('call the "removeShare"-action on click', async () => {
      const { wrapper } = createWrapper()
      const emitSpy = jest.spyOn(wrapper.vm, '$emit')
      await wrapper.findAll(selectors.removeShareBtn).trigger('click')
      expect(emitSpy).toHaveBeenCalledWith('removeShare')
    })
  })
  describe('expiration date', () => {
    describe('set-expiration button', () => {
      it.each(['user', 'group'])('renders the button', (shareType) => {
        const { wrapper } = createWrapper({ props: { shareCategory: shareType } })
        expect(wrapper.find(selectors.findCollaboratorsExpirationBtn).exists()).toBeTruthy()
      })
    })
    describe('remove-expiration button', () => {
      it('renders the button', () => {
        const { wrapper } = createWrapper({
          props: { shareCategory: 'user', expirationDate: new Date() }
        })
        expect(wrapper.find(selectors.removeExpirationDateBtn).exists()).toBeTruthy()
      })
      it('calls the "removeExpirationDate"-action on click', async () => {
        const { wrapper } = createWrapper({
          props: { shareCategory: 'user', expirationDate: new Date() }
        })
        const emitSpy = jest.spyOn(wrapper.vm, '$emit')
        wrapper.vm.$refs.expirationDateDrop = mock<Element>()
        wrapper.find(selectors.removeExpirationDateBtn).trigger('click')
        expect(emitSpy).toHaveBeenCalledWith('expirationDateChanged', { expirationDate: null })
      })
    })
    describe('max expiration date', () => {
      it.each(['user', 'group'])(
        'gets passed correctly to the datepicker component',
        (shareType) => {
          const { wrapper } = createWrapper({
            stubs: { 'oc-datepicker': true },
            props: { shareCategory: shareType, expirationDate: new Date() },
            expirationDateEnforced: true
          })
          expect(wrapper.find(selectors.datepickerStub).attributes()['max-date']).toBeDefined()
        }
      )
      it.each(['user', 'group'])(
        'does not get passed to the datepicker component when expire date is disabled globally',
        (shareType) => {
          const { wrapper } = createWrapper({
            stubs: { 'oc-datepicker': true },
            props: { shareCategory: shareType, expirationDate: new Date() },
            expirationDateEnforced: true,
            expireDateEnabled: false
          })
          expect(wrapper.find(selectors.datepickerStub).attributes()['max-date']).toBeUndefined()
        }
      )
    })
  })
})

function createWrapper({
  stubs = {},
  props = {},
  expirationDateEnforced = false,
  expireDateEnabled = true
} = {}) {
  const storeOptions = defaultStoreMockOptions
  storeOptions.getters.capabilities.mockImplementation(() => ({
    files_sharing: {
      user: {
        expire_date: { enabled: expireDateEnabled, enforced: expirationDateEnforced, days: 7 }
      },
      group: {
        expire_date: { enabled: expireDateEnabled, enforced: expirationDateEnforced, days: 7 }
      }
    }
  }))
  const store = createStore(storeOptions)
  return {
    wrapper: mount(EditDropdown, {
      props: { canEditOrDelete: true, ...props },
      global: {
        plugins: [...defaultPlugins(), store],
        mocks: defaultComponentMocks(),
        stubs: { ...stubs }
      }
    })
  }
}
