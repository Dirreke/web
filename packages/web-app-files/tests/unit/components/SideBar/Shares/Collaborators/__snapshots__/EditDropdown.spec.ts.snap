// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditDropdown default actions render "remove share" and "access details" as default 1`] = `
<span class="oc-flex oc-flex-middle"><button type="button" class="collaborator-edit-dropdown-options-btn oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw" id="files-collaborators-edit-button-5"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span></button>
<div id="oc-drop-4" class="oc-drop oc-box-shadow-medium oc-rounded">
  <div class="oc-card oc-card-body oc-background-secondary oc-p-s">
    <ul aria-label="Context menu of the share" class="oc-list oc-my-rm oc-mx-rm collaborator-edit-dropdown-options-list">
      <!---->
      <li class="oc-rounded oc-menu-item-hover"><button type="button" class="oc-p-s action-menu-item oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw remove-share" data-testid="collaborator-remove-share-btn"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span> <span>Remove share</span></button></li>
      <li class="oc-rounded oc-menu-item-hover"><button type="button" class="oc-p-s action-menu-item oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw show-access-details"><span class="oc-icon oc-icon-m oc-icon-passive"><!----></span> <span>Access details</span></button></li>
    </ul>
  </div>
</div>
</span>
`;
